File: examples/control/RelayOscillator/RelayOscillator.cpp
/* Siconos is a program dedicated to modeling, simulation and control
 * of non smooth dynamical systems.
 *
 * Copyright 2023 INRIA.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <SolverOptions.h>

#include <SiconosKernel.hpp>
#include <chrono>
#include <string>
using namespace std;
using Matrix = siconos::algebra::SiconosMatrix;
using Vector = siconos::algebra::SiconosVector;

// main program
/* Example of a limit cycle in Relay system predicted by the describing function approach
 * see Example 7.9 in "Nonlinear Systems" H. Khalil. Third Edition. Prentice Hall, 2002 ISBN
 * 0-13-067389-7
 *
 * The transfer function of the linear part is $G(s) 1/(s(s+1)(s+2))$ and the describing
 * function is $\Psi(a)=4/(\pi a)$ The predicted limit cycle is for the output $y(t) = a \sin
 * (\omega t)$ with $a = 2/(3\pi)$ and $\omega = \sqrt(2)$.
 */

int main(int argc, char* argv[]) {
  // Exception handling
  try {
    // == User-defined parameters ==
    unsigned int ndof = 3;  // number of degrees of freedom of your system
    double t0 = 0.0;
    double T = 100;     // Total simulation times
    double h = 1.0e-2;  // Time step
    double xinit = 3.0 * sqrt(2.0) / (2.0 * M_PI);

    // ================= Creation of the model =======================
    // Steps:
    // - create some Dynamical Systems
    // - create some Interactions between those Dynamical Systems
    //   Interaction = some relations (constraints) and a NonSmoothLaw
    // - create a NonSmoothDynamicalSystem with the DynamicalSystems and the Interactions
    // - add this NonSmoothDynamicalSystem into a Model
    // - add a Simulation to the model
    //  Simulation = TimeDiscretisation + OneStepIntegrator and OneStepNSProblem

    // -------------------------
    // --- Dynamical systems ---
    // -------------------------

    // First System:
    // dx/dt = Ax + u(t) + r
    // x(0) = x0
    // Note: r = Blambda, B defines in relation below.

    auto A = std::make_shared<Matrix>(ndof, ndof);
    (*A)(0, 0) = 0.0;
    (*A)(0, 1) = 1.0;
    (*A)(0, 2) = 0.0;
    (*A)(1, 0) = 0.0;
    (*A)(1, 1) = 0.0;
    (*A)(1, 2) = 1.0;
    (*A)(2, 0) = 0.0;
    (*A)(2, 1) = -3.0;
    (*A)(2, 2) = -2.0;
    auto x0 = std::make_shared<Vector>(ndof);
    (*x0)(0) = 0.0;
    (*x0)(1) = xinit;
    (*x0)(2) = 0.0;

    auto process = std::make_shared<siconos::modeling::FirstOrderLinearDS>(*x0, *A);
    //    process->setComputebFunction("ObserverLCSPlugin","uProcess");

    // --------------------
    // --- Interactions ---
    // --------------------
    unsigned int ninter = 1;  // dimension of your Interaction = size of y and lambda vectors

    // First relation, related to the process
    // y = Cx + Dlambda
    // r = Blambda
    auto B = std::make_shared<Matrix>(ndof, ninter);
    (*B)(0, 0) = 0.0;
    (*B)(1, 0) = 0.0;
    (*B)(2, 0) = 1.0;

    auto C = std::make_shared<Matrix>(ninter, ndof);
    (*C)(0, 0) = 1.0;
    (*C)(0, 1) = 0.0;
    (*C)(0, 2) = 0.0;

    auto myProcessRelation = std::make_shared<siconos::modeling::FirstOrderLinearR>(C, B);
    auto D = std::make_shared<Matrix>(ninter, ninter);
    (*D)(0, 0) = 0.0;

    myProcessRelation->setConstantD(*D);
    // myProcessRelation->setComputeEFunction("ObserverLCSPlugin","computeE");

    // Second relation, related to the observer
    // haty = C hatX + D hatLambda + E
    // hatR = B hatLambda

    // NonSmoothLaw
    unsigned int nslawSize = 1;
    auto myNslaw = std::make_shared<siconos::modeling::RelayNSL>(nslawSize);

    myNslaw->display();

    // The Interaction which involves the first DS (the process)
    auto myProcessInteraction =
        std::make_shared<siconos::modeling::Interaction>(myNslaw, myProcessRelation);

    // -------------
    // --- Model ---
    // -------------
    auto relayOscillator =
        std::make_shared<siconos::modeling::NonSmoothDynamicalSystem>(t0, T);
    relayOscillator->insertDynamicalSystem(process);
    relayOscillator->link(myProcessInteraction, process);

    // ------------------
    // --- Simulation ---
    // ------------------
    // TimeDiscretisation
    auto td = std::make_shared<siconos::simulation::TimeDiscretisation>(t0, h);
    // == Creation of the Simulation ==
    auto s = std::make_shared<siconos::simulation::TimeStepping>(relayOscillator, td);
    // -- OneStepIntegrators --
    double theta = 0.5;
    auto myIntegrator = std::make_shared<siconos::integrators::EulerMoreauOSI>(theta);
    s->insertIntegrator(myIntegrator);

    // -- OneStepNsProblem --

    auto osnspb = std::make_shared<siconos::nonsmooth_formulations::Relay>();

    osnspb->setSolverId(SICONOS_RELAY_LEMKE);
    osnspb->numericsSolverOptions()->dparam[0] = 1e-08;
    s->insertNonSmoothProblem(osnspb);

    // =========================== End of model definition ===========================

    // ================================= Computation =================================

    // --- Get the values to be plotted ---
    unsigned int outputSize = 7;          // number of required data
    unsigned int N = ceil((T - t0) / h);  // Number of time steps

    Matrix dataPlot(N, outputSize);

    auto xProc = process->x();
    auto lambdaProc = myProcessInteraction->lambda(0);
    auto yProc = myProcessInteraction->y(0);
    unsigned int k = 0;  // Current step

    // -> saved in a matrix dataPlot
    dataPlot(0, 0) = relayOscillator->t0();  // Initial time of the model
    dataPlot(k, 1) = (*xProc)(0);
    dataPlot(k, 2) = (*xProc)(1);
    dataPlot(k, 3) = (*xProc)(2);
    dataPlot(k, 4) = (*lambdaProc)(0);
    dataPlot(k, 5) = (*yProc)(0);

    // ==== Simulation loop =====
    cout << "====> Start computation ... " << endl << endl;

    // *z = *(myProcessInteraction->y(0)->getVectorPtr(0));
    // Simulation loop
    std::chrono::time_point<std::chrono::system_clock> start, end;
    start = std::chrono::system_clock::now();
    while (k < N - 1) {
      k++;

      //  osnspb->setNumericsVerboseMode(1);

      //  *z = *(myProcessInteraction->y(0)->getVectorPtr(0));
      s->computeOneStep();
      dataPlot(k, 0) = s->nextTime();
      dataPlot(k, 1) = (*xProc)(0);
      dataPlot(k, 2) = (*xProc)(1);
      dataPlot(k, 3) = (*xProc)(2);
      dataPlot(k, 4) = (*lambdaProc)(0);
      dataPlot(k, 5) = (*yProc)(0);
      s->nextStep();
    }
    cout << endl << "End of computation - Number of iterations done: " << k - 1 << endl;
    cout << "Computation Time \n";
    ;
    end = std::chrono::system_clock::now();
    auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count();
    cout << "Computation time : " << elapsed << " ms\n";
    // --- Output files ---
    cout << "====> Output file writing ...\n";

    siconos::algebra::io::write("RelayOscillator.dat", dataPlot,
                                siconos::algebra::io::ASCII_OUT,
                                siconos::algebra::io::WriteType::nodim);

    double error = 0.0, eps = 1e-12;
    if ((error = siconos::algebra::io::compareRefFile(dataPlot, "RelayOscillator.ref", eps)) >
        eps)
      return 1;
    else
      return 0;

  }

  catch (...) {
    siconos::exception::process();
    return 1;
  }
}

File: examples/control/RelayOscillator/RelayOscillator.py
#!/usr/bin/env python

# Siconos is a program dedicated to modeling, simulation and control
# of non smooth dynamical systems.
#
# Copyright 2021 INRIA.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
import math
import numpy as np
from siconos.modeling import (
    FirstOrderLinearDS,
    FirstOrderLinearTIR,
    RelayNSL,
    Interaction,
    NonSmoothDynamicalSystem,
)

import siconos.integrators

import siconos.plot_config as sicoplot

# Turn off interactive backend by default
plt, enable_plot = sicoplot.choose_backend(False)

t0 = 0.0
T = 100  # Total simulation time
h_step = 1.0e-2  # Time step
xinit = 3.0 * math.sqrt(2.0) / (2.0 * math.pi)  # initial voltage
Modeltitle = "RelayOscillator"


#
# dynamical system
#
init_state = [0.0, xinit, 0]

A = [[0, 1.0, 0.0], [0.0, 0.0, 1.0], [0.0, -3.0, -2.0]]

LSRelayOscillator = FirstOrderLinearDS(init_state, A)

#
# Interactions
#

C = [[1.0, 0.0, 0.0]]

D = [[0.0]]

B = [[0.0], [0.0], [1.0]]

LTIRRelayOscillator = FirstOrderLinearTIR(C, B)
LTIRRelayOscillator.setConstantD(D)

nslaw = RelayNSL(1)
InterRelayOscillator = Interaction(nslaw, LTIRRelayOscillator)


#
# Model
#
relayOscillator = NonSmoothDynamicalSystem(t0, T)
relayOscillator.setTitle(Modeltitle)
#   add the dynamical system in the non smooth dynamical system
relayOscillator.insertDynamicalSystem(LSRelayOscillator)


#   link the interaction and the dynamical system
relayOscillator.link(InterRelayOscillator, LSRelayOscillator)


#
# Simulation
#

# (1) OneStepIntegrators
theta = 0.5
aOSI = siconos.integrators.EulerMoreauOSI(theta)
# (2) Time discretisation
aTiDisc = siconos.simulation.TimeDiscretisation(t0, h_step)

# (3) Non smooth problem
aRelay = siconos.nonsmooth_formulations.Relay()

# (4) Simulation setup with (1) (2) (3)
aTS = siconos.simulation.TimeStepping(relayOscillator, aTiDisc, aOSI, aRelay)

# end of model definition

#
# computation
#


k = 0
h = aTS.timeStep()
print("Timestep : ", h)
# Number of time steps
N = (int)((T - t0) / h)
print("Number of steps : ", N)

# Get the values to be plotted
# ->saved in a matrix dataPlot

dataPlot = np.empty([N + 1, 8])

x = LSRelayOscillator.x()
print("Initial state : ", x)
y = InterRelayOscillator.y(0)
print("First y : ", y)
lambda_ = InterRelayOscillator.lambda_(0)

while k < N:
    aTS.computeOneStep()
    # aLCP.display()
    dataPlot[k, 0] = aTS.nextTime()
    #  inductor voltage
    dataPlot[k, 1] = x[0]
    dataPlot[k, 2] = x[1]
    dataPlot[k, 3] = x[2]
    dataPlot[k, 4] = y[0]
    dataPlot[k, 5] = lambda_[0]

    k += 1
    if k % 1000 == 0:
        print("step =", k, " < ", N)
    aTS.nextStep()

if enable_plot:
    #
    # plots
    #
    plt.subplot(511)
    plt.title("x1")
    plt.plot(dataPlot[0 : k - 1, 0], dataPlot[0 : k - 1, 1])
    plt.grid()
    plt.subplot(512)
    plt.title("x2")
    plt.plot(dataPlot[0 : k - 1, 0], dataPlot[0 : k - 1, 2])
    plt.grid()
    plt.subplot(513)
    plt.title("x3")
    plt.plot(dataPlot[0 : k - 1, 0], dataPlot[0 : k - 1, 3])
    plt.subplot(514)
    plt.title("y")
    plt.plot(dataPlot[0 : k - 1, 0], dataPlot[0 : k - 1, 4])
    plt.grid()
    plt.subplot(515)
    plt.title("lambda")
    plt.plot(dataPlot[0 : k - 1, 0], dataPlot[0 : k - 1, 5])
    plt.grid()
    plt.savefig("relay_oscillator.png")
